{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "Collapsed": "false"
   },
   "outputs": [],
   "source": [
    "# Import libraries\n",
    "import requests\n",
    "import os\n",
    "from time import sleep\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "Collapsed": "false"
   },
   "outputs": [],
   "source": [
    "#------------------------------------------------------------------------------\n",
    "#   Section with general functions\n",
    "#------------------------------------------------------------------------------\n",
    "def download(url, fileName):\n",
    "    for i in range(10):\n",
    "        try:\n",
    "            # Delete existing files with filename\n",
    "            try:\n",
    "                os.remove(fileName) \n",
    "            except:\n",
    "                pass\n",
    "            \n",
    "            \"\"\" Use requests to download file. \n",
    "            Works with streams to be able large files without having the need of a \n",
    "            large memory.\n",
    "            \"\"\"\n",
    "            with requests.get(url, stream=True) as r:\n",
    "                r.raise_for_status()\n",
    "                with open(fileName, 'wb') as f:\n",
    "                    for chunk in r.iter_content(chunk_size=8192): \n",
    "                        if chunk:\n",
    "                            f.write(chunk)\n",
    "            return fileName\n",
    "        except:\n",
    "            print(\"Download\", url,\"failed:\",i)\n",
    "            sleep(5)\n",
    "            \n",
    "def uniprotRetrieve(fileName, query=\"\",format=\"list\",columns=\"\",include=\"no\",compress=\"no\",limit=0,offset=0):\n",
    "    \"\"\"Downloads file from uniprot for given parameters\n",
    "    \n",
    "    If no parameters are given the function will download a list of all the \n",
    "    proteins ID's. More information about how the URL should be constructed can\n",
    "    be found on: \n",
    "    https://www.uniprot.org/help/api%5Fqueries\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    fileName : str\n",
    "        name for the downloaded file\n",
    "    query : str (Default='')\n",
    "        query that would be searched if as you used the webinterface on \n",
    "        https://www.uniprot.org/. If no query is provided, all protein entries\n",
    "        are selected. \n",
    "    format : str (Default='list')\n",
    "        File format you want to retrieve from uniprot. Available format are:\n",
    "        html | tab | xls | fasta | gff | txt | xml | rdf | list | rss\n",
    "    columns : str (Default='')\n",
    "        Column information you want to know for each entry in the query \n",
    "        when format tab or xls is selected.\n",
    "    include : str (Default='no')\n",
    "        Include isoform sequences when the format parameter is set to fasta.\n",
    "        Include description of referenced data when the format parameter is set to rdf.\n",
    "        This parameter is ignored for all other values of the format parameter.\n",
    "    compress : str (Default='no')\n",
    "        download file in gzipped compression format.\n",
    "    limit : int (Default=0)\n",
    "        Limit the amount of results that is given. 0 means you download all.\n",
    "    offset : int (Default=0)\n",
    "        When you limit the amount of results, offset determines where to start.\n",
    "        \n",
    "    Returns\n",
    "    -------\n",
    "    fileName : str\n",
    "        Name of the downloaeded file.\n",
    "    \"\"\"\n",
    "    def generateURL(baseURL, query=\"\",format=\"list\",columns=\"\",include=\"no\",compress=\"no\",limit=\"0\",offset=\"0\"):\n",
    "        \"\"\"Generate URL with given parameters\"\"\"\n",
    "        def glueParameters(**kwargs):\n",
    "            gluedParameters = \"\"\n",
    "            for parameter, value in kwargs.items():\n",
    "                gluedParameters+=parameter + \"=\" + str(value) + \"&\"\n",
    "            return gluedParameters.replace(\" \",\"+\")[:-1] #Last \"&\" is removed, spacec replaced by \"+\"\n",
    "        return baseURL + glueParameters(query=query,\n",
    "                                        format=format,\n",
    "                                        columns=columns,\n",
    "                                        include=include,\n",
    "                                        compress=compress,\n",
    "                                        limit=limit,\n",
    "                                        offset=offset)\n",
    "    URL = generateURL(\"https://www.uniprot.org/uniprot/?\",\n",
    "               query=query,\n",
    "               format=format,\n",
    "               columns=columns,\n",
    "               include=include,\n",
    "               compress=compress,\n",
    "               limit=limit,\n",
    "               offset=offset)\n",
    "    return download(URL, fileName)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "Collapsed": "false"
   },
   "outputs": [],
   "source": [
    "# Download organisms of all proteins in gammaproteobacteria \n",
    "QUERY='taxonomy:gammaproteobacteria'\n",
    "FORMAT=\"tab\"\n",
    "COLUMNS=\"organism,lineage(SPECIES),lineage(GENUS)\"\n",
    "FILENAME=\"gammaproteobacteriaOrganisms.tab\"\n",
    "FILENAME = uniprotRetrieve(FILENAME,query=QUERY,format=FORMAT,columns=COLUMNS)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "Collapsed": "false"
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Organism</th>\n",
       "      <th>Taxonomic lineage (SPECIES)</th>\n",
       "      <th>Taxonomic lineage (GENUS)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>Xanthomonas campestris pv. campestris (strain ...</td>\n",
       "      <td>Xanthomonas campestris</td>\n",
       "      <td>Xanthomonas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>Alkalilimnicola ehrlichii (strain ATCC BAA-110...</td>\n",
       "      <td>Alkalilimnicola ehrlichii</td>\n",
       "      <td>Alkalilimnicola</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>Xanthomonas oryzae pv. oryzae (strain PXO99A)</td>\n",
       "      <td>Xanthomonas oryzae</td>\n",
       "      <td>Xanthomonas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>Shewanella pealeana (strain ATCC 700345 / ANG-...</td>\n",
       "      <td>Shewanella pealeana</td>\n",
       "      <td>Shewanella</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>Hydrogenovibrio crunogenus (strain XCL-2) (Thi...</td>\n",
       "      <td>Hydrogenovibrio crunogenus</td>\n",
       "      <td>Hydrogenovibrio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>27597072</td>\n",
       "      <td>Brenneria alni</td>\n",
       "      <td>Brenneria alni</td>\n",
       "      <td>Brenneria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>27597073</td>\n",
       "      <td>Alcanivorax sp. DG881</td>\n",
       "      <td>Alcanivorax sp. DG881</td>\n",
       "      <td>Alcanivorax</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>27597074</td>\n",
       "      <td>Salmonella typhimurium (strain SL1344)</td>\n",
       "      <td>Salmonella enterica (Salmonella choleraesuis)</td>\n",
       "      <td>Salmonella</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>27597075</td>\n",
       "      <td>Pseudomonas soli</td>\n",
       "      <td>Pseudomonas soli</td>\n",
       "      <td>Pseudomonas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>27597076</td>\n",
       "      <td>Pseudomonas nitroreducens</td>\n",
       "      <td>Pseudomonas nitroreducens</td>\n",
       "      <td>Pseudomonas</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>27597077 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                   Organism  \\\n",
       "0         Xanthomonas campestris pv. campestris (strain ...   \n",
       "1         Alkalilimnicola ehrlichii (strain ATCC BAA-110...   \n",
       "2             Xanthomonas oryzae pv. oryzae (strain PXO99A)   \n",
       "3         Shewanella pealeana (strain ATCC 700345 / ANG-...   \n",
       "4         Hydrogenovibrio crunogenus (strain XCL-2) (Thi...   \n",
       "...                                                     ...   \n",
       "27597072                                     Brenneria alni   \n",
       "27597073                              Alcanivorax sp. DG881   \n",
       "27597074             Salmonella typhimurium (strain SL1344)   \n",
       "27597075                                   Pseudomonas soli   \n",
       "27597076                          Pseudomonas nitroreducens   \n",
       "\n",
       "                            Taxonomic lineage (SPECIES)  \\\n",
       "0                                Xanthomonas campestris   \n",
       "1                             Alkalilimnicola ehrlichii   \n",
       "2                                    Xanthomonas oryzae   \n",
       "3                                   Shewanella pealeana   \n",
       "4                            Hydrogenovibrio crunogenus   \n",
       "...                                                 ...   \n",
       "27597072                                 Brenneria alni   \n",
       "27597073                          Alcanivorax sp. DG881   \n",
       "27597074  Salmonella enterica (Salmonella choleraesuis)   \n",
       "27597075                               Pseudomonas soli   \n",
       "27597076                      Pseudomonas nitroreducens   \n",
       "\n",
       "         Taxonomic lineage (GENUS)  \n",
       "0                      Xanthomonas  \n",
       "1                  Alkalilimnicola  \n",
       "2                      Xanthomonas  \n",
       "3                       Shewanella  \n",
       "4                  Hydrogenovibrio  \n",
       "...                            ...  \n",
       "27597072                 Brenneria  \n",
       "27597073               Alcanivorax  \n",
       "27597074                Salmonella  \n",
       "27597075               Pseudomonas  \n",
       "27597076               Pseudomonas  \n",
       "\n",
       "[27597077 rows x 3 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read in proteins\n",
    "FILENAME=\"gammaproteobacteriaOrganisms.tab.bu\"\n",
    "PROTEINS= pd.read_csv(FILENAME,sep=\"\\t\")\n",
    "PROTEINS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "Collapsed": "false"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "25663 Organism\n",
      "20950 Taxonomic lineage (SPECIES)\n",
      "455 Taxonomic lineage (GENUS)\n"
     ]
    }
   ],
   "source": [
    "# See how many unique units there are at different taxonomic levels\n",
    "for LVL in [\"Organism\", \"Taxonomic lineage (SPECIES)\", \"Taxonomic lineage (GENUS)\"]:\n",
    "    print(len(PROTEINS[LVL].unique()),LVL)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "Collapsed": "false"
   },
   "outputs": [],
   "source": [
    "#uniqueOrganisms = proteins[\"Organism\"].unique()\n",
    "#organisms = pd.DataFrame(uniqueOrganisms,columns=[\"Organism\"])\n",
    "#organisms.to_csv(\"organisms.tab\",sep=\"\\t\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "Collapsed": "false"
   },
   "outputs": [],
   "source": [
    "# Download organisms of all proteins in gammaproteobacteria \n",
    "QUERY='taxonomy:Enterobacterales'\n",
    "FORMAT=\"tab\"\n",
    "COLUMNS=\"organism,lineage(SPECIES),lineage(GENUS)\"\n",
    "FILENAME=\"enterobacteralesProteins.tab\"\n",
    "FILENAME = uniprotRetrieve(FILENAME,query=QUERY,format=FORMAT,columns=COLUMNS)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "Collapsed": "false"
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Organism</th>\n",
       "      <th>Taxonomic lineage (SPECIES)</th>\n",
       "      <th>Taxonomic lineage (GENUS)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>Shigella boydii serotype 4 (strain Sb227)</td>\n",
       "      <td>Shigella boydii</td>\n",
       "      <td>Shigella</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>Yersinia pestis bv. Antiqua (strain Antiqua)</td>\n",
       "      <td>Yersinia pestis</td>\n",
       "      <td>Yersinia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>Buchnera aphidicola subsp. Baizongia pistaciae...</td>\n",
       "      <td>Buchnera aphidicola</td>\n",
       "      <td>Buchnera (aphid P-endosymbionts)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>Cronobacter sakazakii (strain ATCC BAA-894) (E...</td>\n",
       "      <td>Cronobacter sakazakii (Enterobacter sakazakii)</td>\n",
       "      <td>Cronobacter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>Klebsiella pneumoniae (strain 342)</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>Klebsiella</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8878902</td>\n",
       "      <td>Shigella boydii serotype 18 (strain CDC 3083-9...</td>\n",
       "      <td>Shigella boydii</td>\n",
       "      <td>Shigella</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8878903</td>\n",
       "      <td>Yokenella regensburgei ATCC 43003</td>\n",
       "      <td>Yokenella regensburgei</td>\n",
       "      <td>Yokenella</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8878904</td>\n",
       "      <td>Serratia sp. DD3</td>\n",
       "      <td>Serratia sp. DD3</td>\n",
       "      <td>Serratia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8878905</td>\n",
       "      <td>Brenneria alni</td>\n",
       "      <td>Brenneria alni</td>\n",
       "      <td>Brenneria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8878906</td>\n",
       "      <td>Salmonella typhimurium (strain SL1344)</td>\n",
       "      <td>Salmonella enterica (Salmonella choleraesuis)</td>\n",
       "      <td>Salmonella</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8878907 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  Organism  \\\n",
       "0                Shigella boydii serotype 4 (strain Sb227)   \n",
       "1             Yersinia pestis bv. Antiqua (strain Antiqua)   \n",
       "2        Buchnera aphidicola subsp. Baizongia pistaciae...   \n",
       "3        Cronobacter sakazakii (strain ATCC BAA-894) (E...   \n",
       "4                       Klebsiella pneumoniae (strain 342)   \n",
       "...                                                    ...   \n",
       "8878902  Shigella boydii serotype 18 (strain CDC 3083-9...   \n",
       "8878903                  Yokenella regensburgei ATCC 43003   \n",
       "8878904                                   Serratia sp. DD3   \n",
       "8878905                                     Brenneria alni   \n",
       "8878906             Salmonella typhimurium (strain SL1344)   \n",
       "\n",
       "                            Taxonomic lineage (SPECIES)  \\\n",
       "0                                       Shigella boydii   \n",
       "1                                       Yersinia pestis   \n",
       "2                                   Buchnera aphidicola   \n",
       "3        Cronobacter sakazakii (Enterobacter sakazakii)   \n",
       "4                                 Klebsiella pneumoniae   \n",
       "...                                                 ...   \n",
       "8878902                                 Shigella boydii   \n",
       "8878903                          Yokenella regensburgei   \n",
       "8878904                                Serratia sp. DD3   \n",
       "8878905                                  Brenneria alni   \n",
       "8878906   Salmonella enterica (Salmonella choleraesuis)   \n",
       "\n",
       "                Taxonomic lineage (GENUS)  \n",
       "0                                Shigella  \n",
       "1                                Yersinia  \n",
       "2        Buchnera (aphid P-endosymbionts)  \n",
       "3                             Cronobacter  \n",
       "4                              Klebsiella  \n",
       "...                                   ...  \n",
       "8878902                          Shigella  \n",
       "8878903                         Yokenella  \n",
       "8878904                          Serratia  \n",
       "8878905                         Brenneria  \n",
       "8878906                        Salmonella  \n",
       "\n",
       "[8878907 rows x 3 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read in proteins\n",
    "FILENAME=\"enterobacteralesProteins.tab.bu\"\n",
    "PROTEINS= pd.read_csv(FILENAME,sep=\"\\t\")\n",
    "PROTEINS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "Collapsed": "false"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6127 Organism\n",
      "3102 Taxonomic lineage (SPECIES)\n",
      "99 Taxonomic lineage (GENUS)\n"
     ]
    }
   ],
   "source": [
    "# See how many unique units there are at different taxonomic levels\n",
    "for LVL in [\"Organism\", \"Taxonomic lineage (SPECIES)\", \"Taxonomic lineage (GENUS)\"]:\n",
    "    print(len(PROTEINS[LVL].unique()),LVL)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "Collapsed": "false"
   },
   "outputs": [],
   "source": [
    "uniqueOrganisms = PROTEINS[\"Taxonomic lineage (SPECIES)\"].unique()\n",
    "organisms = pd.DataFrame(uniqueOrganisms,columns=[\"Organism\"])\n",
    "organisms.to_csv(\"organisms.tab\",sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "Collapsed": "false"
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
